File: C++/shortest-completing-word.cpp
// Time:  O(n)
// Space: O(1)

class Solution {
public:
    string shortestCompletingWord(string licensePlate, vector<string>& words) {
        string result;
        const auto& counter = counts(licensePlate);
        for (const auto& word : words) {
            if ((result.empty() || word.length() < result.length()) &&
               contains(counter, word)) {
                result = word;
            }
        }
        return result;
    }
    
private:
    bool contains(const vector<int>& counter1, const string& w2) const {
        const auto& counter2 = counts(w2);
        for (int i = 0; i < counter2.size(); ++i) {
            if (counter1[i] > counter2[i]) {
                return false;
            }
        }
        return true;
    }
    
    vector<int> counts(const string& s) const {
        vector<int> count(26);
        for (const auto& c : s) {
            if (isalpha(c)) {
                ++count[tolower(c) - 'a'];
            }
        }
        return count;
    }
};

File: Python/shortest-completing-word.py
# Time:  O(n)
# Space: O(1)

import collections


class Solution(object):
    def shortestCompletingWord(self, licensePlate, words):
        """
        :type licensePlate: str
        :type words: List[str]
        :rtype: str
        """
        def contains(counter1, w2):
            c2 = collections.Counter(w2.lower())
            c2.subtract(counter1)
            return all(map(lambda x: x >= 0, c2.values()))

        result = None
        counter = collections.Counter(c.lower() for c in licensePlate if c.isalpha())
        for word in words:
            if (result is None or (len(word) < len(result))) and \
               contains(counter, word):
                result = word
        return result

